### PR TITLE
feat(timers): a stab at fake timers

### PR DESCRIPTION
**Pull request by pavelfeldman**

### CODE REVIEW

File: docs/src/api/class-clock.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+* since: v1.45
+- `now` <[int]|[Date]>
+
+Install fake timers with the specified unix epoch (default: 0).
+
+### option: Clock.install.toFake
+* since: v1.45
+- `toFake` <[Array]<[FakeMethod]<"setTimeout"|"clearTimeout"|"setImmediate"|"clearImmediate"|"setInterval"|"clearInterval"|"Date"|"requestAnimationFrame"|"cancelAnimationFrame"|"requestIdleCallback"|"cancelIdleCallback"|"performance">>>
+
+An array with names of global methods and APIs to fake. For instance, `await page.clock.install({ toFake: ['setTimeout'] })` will fake only `setTimeout()`.
```
Comment (dgozman) [inline docs/src/api/class-clock.md:22]: Let's mention the default.

File: docs/src/api/class-clock.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+### option: Clock.install.loopLimit
+* since: v1.45
+- `loopLimit` <[int]>
+
+The maximum number of timers that will be run when calling runAll() (default: 1000).
```
Comment (dgozman) [inline docs/src/api/class-clock.md:28]:
```suggestion
The maximum number of timers that will be run when calling [`method: Clock.runAll`], defaults to `1000`.
```

File: docs/src/api/class-clock.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+### option: Clock.install.shouldAdvanceTime
+* since: v1.45
+- `shouldAdvanceTime` <[boolean]>
+
+Tells @sinonjs/fake-timers to increment mocked time automatically based on the real system time shift (e.g., the mocked time will be incremented by
```
Comment (dgozman) [inline docs/src/api/class-clock.md:34]:
```suggestion
Tells `@sinonjs/fake-timers` to increment mocked time automatically based on the real system time shift (e.g., the mocked time will be incremented by
```

File: docs/src/api/class-clock.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+20ms for every 20ms change in the real system time) (default: false).
```
Comment (dgozman) [inline docs/src/api/class-clock.md:35]:
```suggestion
20ms for every 20ms change in the real system time). Defaults to `false`.
```

File: docs/src/api/class-clock.md [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+### option: Clock.install.advanceTimeDelta
+* since: v1.45
+- `advanceTimeDelta` <[int]>
+
+Relevant only when using with shouldAdvanceTime: true. Increment mocked time by advanceTimeDelta ms every advanceTimeDelta ms change
```
Comment (dgozman) [inline docs/src/api/class-clock.md:41]:
```suggestion
Relevant only when using with [`option: shouldAdvanceTime`]. Increment mocked time by advanceTimeDelta ms every advanceTimeDelta ms change
```

File: docs/src/api/class-clock.md [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+in the real system time (default: 20).
+
+## async method: Clock.jump
```
Comment (dgozman) [inline docs/src/api/class-clock.md:44]: missing `return` type

File: docs/src/api/class-clock.md [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+Time may be the number of milliseconds to advance the clock by or a human-readable string. Valid string formats are "08" for eight seconds, "01:00" for one minute and "02:34:10" for two hours, 34 minutes and ten seconds.
+
+
+## async method: Clock.runAll
+* since: v1.45
+- returns: <[int]> Fake milliseconds since the unix epoch.
+
+Runs all pending timers until there are none remaining. If new timers are added while it is executing they will be run as well.
+This makes it easier to run asynchronous tests to completion without worrying about the number of timers they use, or the delays in those timers.
+It runs a maximum of loopLimit times after which it assumes there is an infinite loop of timers and throws an error.
```
Comment (dgozman) [inline docs/src/api/class-clock.md:63]:
```suggestion
It runs a maximum of [`option: loopLimit`] times after which it assumes there is an infinite loop of timers and throws an error.
```

File: packages/playwright-core/src/server/clock.ts [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+      const module = {};
+      ${fakeTimersSource.source}
+      globalThis.__pwFakeTimers = (module.exports.install())(${JSON.stringify(params)});
+    })();`;
+    await this._addAndEvaluate(script);
+  }
+
+  async jump(time: number | string) {
+    this._assertInstalled();
+    const parsedTime = typeof time === 'number' ? time : parseTime(time);
```
Comment (dgozman) [inline packages/playwright-core/src/server/clock.ts:43]: Just pass `time` to sinon without parsing?

File: packages/playwright-core/src/server/clock.ts [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+
+  async runToLast(): Promise<number> {
+    this._assertInstalled();
+    await this._browserContext.addInitScript(`globalThis.__pwFakeTimers.runToLast()`);
+    return await this._evaluateInFrames(`globalThis.__pwFakeTimers.runToLastAsync()`);
+  }
+
+  async tick(time: number | string): Promise<number> {
+    this._assertInstalled();
+    const parsedTime = typeof time === 'number' ? time : parseTime(time);
```
Comment (dgozman) [inline packages/playwright-core/src/server/clock.ts:61]: Same here, no need to parse.

**Review by dgozman — APPROVED** [review]

File: docs/src/api/class-clock.md (hunk 6 above)
Comment (pavelfeldman) [reply docs/src/api/class-clock.md:44]: Jump does not return.

**Pull request merged**